pyvelox/pyvelox.h

```diff
@@ -19,33 +19,186 @@
-///  for further details.
-inline void addVeloxBindings(pybind11::module& m, bool asLocalModule = true) {
+inline void checkBounds(std::shared_ptr<BaseVector>& v, int64_t idx) {
+  if (idx < 0 || idx >= v->size()) {
+    throw std::out_of_range("Index out of range");
+  }
+}
+
+template <TypeKind T>
+inline auto pyToVariant(const py::handle& obj) {
+  using NativeType = typename TypeTraits<T>::DeepCopiedType;
+  return velox::variant::create<T>(py::cast<NativeType>(obj));
+}
+
+inline velox::variant pyToVariant(const py::handle& obj) {
+  if (obj.is_none()) {
+    return velox::variant();
+  } else if (py::isinstance<py::bool_>(obj)) {
+    return pyToVariant<velox::TypeKind::BOOLEAN>(obj);
+  } else if (py::isinstance<py::int_>(obj)) {
+    return pyToVariant<velox::TypeKind::BIGINT>(obj);
+  } else if (py::isinstance<py::float_>(obj)) {
+    return pyToVariant<velox::TypeKind::DOUBLE>(obj);
+  } else if (py::isinstance<py::str>(obj)) {
+    return pyToVariant<velox::TypeKind::VARCHAR>(obj);
+  } else {
+    throw py::type_error("Invalid type of object");
+  }
+}
+
+template <TypeKind T>
+inline VectorPtr variantsToFlatVector(
+    const std::vector<velox::variant>& variants,
+    facebook::velox::memory::MemoryPool* pool) {
+  using NativeType = typename TypeTraits<T>::NativeType;
+  constexpr bool kNeedsHolder =
+      (T == TypeKind::VARCHAR || T == TypeKind::VARBINARY);
+
+  TypePtr type = fromKindToScalerType(T);
+  auto result =
+      BaseVector::create<FlatVector<NativeType>>(type, variants.size(), pool);
+
+  std::conditional_t<
+      kNeedsHolder,
+      velox::StringViewBufferHolder,
+      velox::memory::MemoryPool*>
+      holder{pool};
+  for (int i = 0; i < variants.size(); i++) {
+    if (variants[i].isNull()) {
+      result->setNull(i, true);
+    } else {
+      if constexpr (kNeedsHolder) {
+        velox::StringView view =
+            holder.getOwnedValue(variants[i].value<std::string>());
+        result->set(i, view);
+      } else {
+        result->set(i, variants[i].value<NativeType>());
+      }
+    }
+  }
+  return result;
+}
+
+inline VectorPtr pyListToVector(
+    const py::list& list,
+    facebook::velox::memory::MemoryPool* pool) {
+  std::vector<velox::variant> variants;
+  variants.reserve(list.size());
+  for (auto item : list) {
+    variants.push_back(pyToVariant(item));
+  }
+
+  if (variants.empty()) {
+    throw py::value_error("Can't create a Velox vector from an empty list");
+  }
+
+  velox::TypeKind first_kind = velox::TypeKind::INVALID;
+  for (velox::variant& var : variants) {
+    if (var.hasValue()) {
+      if (first_kind == velox::TypeKind::INVALID) {
+        first_kind = var.kind();
+      } else if (var.kind() != first_kind) {
+        throw py::type_error(
+            "Velox Vector must consist of items of the same type");
+      }
+    }
+  }
+
+  if (first_kind == velox::TypeKind::INVALID) {
+    throw py::value_error(
+        "Can't create a Velox vector consisting of only None");
```

Comment:
In a separate pr, we can create from_list that takes a type , and an optional list, so we can have empty lists.